security/cmd/chiron/main.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"os"
+	"strings"
+	"time"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/cobra/doc"
+
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+
+	istiocmd "istio.io/istio/pkg/cmd"
+	"istio.io/istio/pkg/kube"
+	"istio.io/istio/security/pkg/cmd"
+	chiron "istio.io/istio/security/pkg/k8s/chiron"
+	"istio.io/pkg/collateral"
+	"istio.io/pkg/ctrlz"
+	"istio.io/pkg/log"
+	"istio.io/pkg/version"
+)
+
+var (
+	opts = cliOptions{
+		logOptions:   log.DefaultOptions(),
+		ctrlzOptions: ctrlz.DefaultOptions(),
+	}
+
+	rootCmd = &cobra.Command{
+		Use:   "Istio Webhook Controller",
+		Short: "Istio Webhook Controller",
+		Args:  cobra.ExactArgs(0),
+		Run: func(cmd *cobra.Command, args []string) {
+			if !opts.enableController {
+				log.Info("Istio Webhook Controller is not enabled, exit")
+				return
+			}
+			runWebhookController()
+		},
+	}
+)
+
+type cliOptions struct {
+	// The namespace of the webhook certificates
+	certificateNamespace string
+	kubeConfigFile       string
+
+	// The file path of k8s CA certificate
+	k8sCaCertFile string
+
+	// The file paths of the mutatingwebhookconfigurations.
+	// In prototype, only one path is supported.
+	mutatingWebhookConfigFiles string
```

Comment:
Do we really need to distinguish MutatingWebhooks and ValidatingWebhooks? Can't we treat them together?

Reply:
MutatingWebhooks and ValidatingWebhooks are distinguished because their types are different, e.g., k8sClient.AdmissionregistrationV1beta1().MutatingWebhookConfigurations() versus k8sClient.AdmissionregistrationV1beta1().ValidatingWebhookConfigurations().